lib/events.js

```diff
@@ -182,55 +180,62 @@ function _addListener(target, type, listener, prepend) {
 
   events = target._events;
   if (events === undefined) {
-    events = target._events = Object.create(null);
-    target._eventsCount = 0;
+    events = target._events = new Map();
   } else {
     // To avoid recursion in the case that type === "newListener"! Before
     // adding it to the listeners, first emit "newListener".
-    if (events.newListener !== undefined) {
-      target.emit('newListener', type,
-                  listener.listener ? listener.listener : listener);
+    if (events.has('newListener')) {
+      target.emit('newListener', type, listener.listener || listener);
 
       // Re-assign `events` because a newListener handler could have caused the
       // this._events to be assigned to a new object
       events = target._events;
+      if (events === undefined)
+        events = target._events = new Map();
     }
-    existing = events[type];
+    existing = events.get(type);
+  }
+
+  if (type === 'error') {
+    events._hasErrorListener = true;
   }
 
   if (existing === undefined) {
     // Optimize the case of one listener. Don't need the extra array object.
-    existing = events[type] = listener;
-    ++target._eventsCount;
+    events.set(type, listener);
+    return target;
+  }
+
+  if (typeof existing === 'function') {
+    // Adding the second element, need to change to array.
+    existing = prepend ? [listener, existing] : [existing, listener];
+    events.set(type, existing);
+  } else if (existing.emitting) {
```

Comment:
Perhaps this check should be more specific to avoid array cloning if we're adding a listener for a different event than what is being emitted.

Reply:
So `existing` actually refers to the specific event so it only clones if we're emitting on that exact event.